product-catalog/app/productcatalog/controllers/ProductDetailPageController.java

```diff
@@ -0,0 +1,161 @@
+                    return getPdpResult(cms, suggestions, productProjection, sku);
+                });
+            });
+        });
+
+        //step 4 recover centrally for exceptions
+        return resultPromise.recover(exception -> {
+            if (exception instanceof ProductNotFoundException || exception instanceof ProductVariantNotFoundException) {
+                return notFoundAction();
+            } else {
+                throw exception;
+            }
+        });
+    }
+
+    //TODO the promise needs to be removed
+    private F.Promise<Result> getPdpResult(final CmsPage cms, final List<ProductProjection> suggestions, final ProductProjection productProjection, final String sku) {
+        final Translator translator = getTranslator();
+        final PriceFormatter priceFormatter = getPriceFormatter();
+        final PriceFinder priceFinder = getPriceFinder();
+
+        final ProductVariant productVariant = obtainProductVariantBySku(sku, productProjection);
+        final ShippingRateDataFactory shippingRateDataFactory = ShippingRateDataFactory.of(priceFormatter);
+        final CategoryLinkDataFactory categoryLinkDataFactory = CategoryLinkDataFactory.of(translator);
+
+        final String additionalTitle = translator.findTranslation(productProjection.getName());
+        final PdpStaticData staticData = new PdpStaticData(cms, BagItemDataFactory.of().create(100), RatingDataFactory.of(cms).create());
+        final List<Category> breadcrumbs = getBreadcrumbsForProduct(productProjection);
+        final List<LinkData> breadcrumbData = breadcrumbs.stream().map(categoryLinkDataFactory::create).collect(toList());
+        final List<ImageData> galleryData = productVariant.getImages().stream().map(ImageData::of).collect(toList());
+        final ProductData productData = ProductDataFactory.of(translator, priceFinder, priceFormatter).create(productProjection, productVariant);
+        final List<ShippingRateData> deliveryData = getShippingRates().stream().map(shippingRateDataFactory::create).collect(toList());
+        final List<ProductThumbnailData> suggestionData = suggestionsToViewData(suggestions);
+
+        final ProductDetailPageContent content = new ProductDetailPageContent(additionalTitle, staticData, breadcrumbData, galleryData, productData, deliveryData, suggestionData);
+
+        return render(view -> ok(view.productDetailPage(content)));
+    }
+
+    private PriceFinder getPriceFinder() {
+        return context().user().priceFinder();
+    }
+
+    private PriceFormatter getPriceFormatter() {
+        return context().user().priceFormatter();
+    }
+
+    private Translator getTranslator() {
+        return context().translator();
+    }
+
+    private ProductVariant obtainProductVariantBySku(final String sku, final ProductProjection productProjection) {
+        final Optional<ProductVariant> variantOptional = productService.findVariantBySku(productProjection, sku);
+        if (variantOptional.isPresent()) {
```

Comment:
I will replace this with a call to https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElseThrow-java.util.function.Supplier-